### PR TITLE
[WIP] New renderer

### LINKED ISSUE #264: Create file watching subsystem
Currently, the inotify support is pretty proof-of-concept only. It's directly integrated into the assetmanager, which should be abstracted in the future.

I'm thinking of creating a `cpp/watch/` subsystem which provides some sane api for registering a watch on a folder, file, etc with requested events.

When the watched events happen, generic callback functions should be called so we can use the watches for any file watching makes sense.

In the background, the "sane api" calls the plattform-specific kernel api, but openage can use just one uniform interface. This way we can support OS X, BSD, etc.

### PR DESCRIPTION
**Pull request by TheJJ**
This is basically a totally new graphics engine, designed for our needs.
- [X] Abstraction of render backend
- [X] OpenGL 3.3 requirement
- [X] Preparations for Vulkan (we don't know much..)
- [X] File watching subsystem
- [X] `constexpr` key-value map
- [X] Screenshot support
  - [ ] optimizations (maybe [this](http://www.songho.ca/opengl/gl_pbo.html#pack) or [this](http://www.opengl-tutorial.org/intermediate-tutorials/tutorial-14-render-to-texture/))
- [ ] Render pipeline abstraction
  - [X] Textures
  - [X] Shaders
  - [x] Compiletime shader variable access
    - [x] Uniforms
    - [x] Vertex attributes
    - [x] Dynamic buffer uploading
  - [ ] Samplers
  - [ ] Materials
  - [ ] Geometries
- [x] Quaternions
- [ ] Texture position transformations without reuploading vertex buffer
- [ ] Texture packing
  - [ ] Efficient binpacking
  - [ ] 3D Textures
- [ ] Separate renderer thread and other event loop improvements
- [ ] Efficient render task queuing (minimize shader/texture changes)
  - [ ] Viewport clipping
- [ ] Smooth zooming
- [ ] Enable Z-Buffering
- [ ] Geometry drawing
  - [ ] squares
  - [ ] circles
  - [ ] smooth paths
  - [ ] boxes (for the gui)
  - [ ] text rendering (use freetype / harfbuzz / pango)
- [ ] Perspective translation (e.g. the unit-selected-circle)
- [ ] Terrain rendering
  - [ ] Terrain texture merging to single square tile
  - [ ] Perspective rendering
  - [ ] Blending cache
  - [ ] Terrain blending in the shader (in perspective to easily split tiles)
- [ ] Outline rendering with obstruction dual-pass (z-flip)
- [ ] [Occlusion](https://a248.e.akamai.net/f/248/10/10/http.developer.nvidia.com/GPUGems/gpugems_ch29.html) [queries](http://vertostudio.com/gamedev/?p=177)
- [ ] Pixel-perfect unit hitbox for unit selection and damage areas

implements #34, #286 and ideas from #153 
closes #264 (file watching subsystem).
will do #149 (terrain blending in shader), #158 /  #154 (blending cache), #141 (terrain perspective adjustments), #359 (outlines), #368 (unit selection box), #374 (texture clipping incorrect).

If you wanna work on this as well, submit pull requests to [my branch](https://github.com/TheJJ/openage/tree/renderer) please!

### CODE REVIEW

**Comment by Jon0** [comment]
can I request a `Texture` constructor which is usable by job threads? Currently the construction leads to gl calls which segfaults for not having an OpenGL context. I have made a small work around by moving the `glBufferData()` call to be done inside the first `Texture::draw()` call of a particular texture. which maybe the best way of implementing this feature. We should also have some kind of check that `Texture::draw()` is called using the main thread.

**Comment by mic-e** [comment]
Maybe we should move all program logic (including the main loop) outside Thread #0 (the thread that owns the OpenGL context and provides input), and have all other threads make synchronous calls into Thread #0?

**Comment by TheJJ** [comment]
My plan was to create a queue structure for all tasks to be rendered. Things to be drawn (textures, texts, geometries, whatever) would be added/registered to a queue and the renderer, when it's ready for the next call, fetches all those instructions, reorders them and then renders them to the framebuffer.

The registering/adding can be done from any thread, the renderer will either fetch the instructions itself or the instructions are converted/processed upon enqueuing, i haven't decided on that yet.

Basically all other subsystems will get backend-independent handles for all the render calls, so opengl or vulcan or maybe html5 someday can do the actual rendering then.

**Comment by TheJJ** [comment]
The thread-independent texture (`TextureData` in `renderer/texture.h`) should now be there i think. You can fill it with data anywhere and have to register it at the renderer instance to actually create the internal opengl handles.

**Comment by franciscod** [comment]
@TheJJ you commented on the roadmap of this PR on IRC, could you write that here?

**Comment by TheJJ** [comment]
Uh, I simply follow the todo list in the top post here, and will actively work on it after https://github.com/SFTtech/kevin is in place to replace travis, no idea how long it'll take.

Occasionaly when I'm motivated to work some thing i'll do it before kevin is there^^

Feel free to contribute to my renderer branch now and submit me pull requests to my account.

**Comment by TheJJ** [comment]
https://software.intel.com/en-us/articles/designing-the-framework-of-a-parallel-game-engine/

**Comment by TheJJ** [comment]
Continued in #850 :)
